### PR TITLE
sstables: introduce HasPartitionTombstones sstable metadata

### PR DESCRIPTION
**Pull request by kbr-**
This is a type with three possible values:
- `Yes` - there are partition tombstones in the sstable
- `No` - there are no partition tombstones
- `Unknown` - we don't know

The metadata is calculated and written for sstable versions greater
or equal to `mc`. For lower versions we don't write it.
When reading, if the metadata is not present, we interpret it as
`Unknown`.

Split from https://github.com/scylladb/scylla/pull/7437.

### CODE REVIEW

File: test/boost/sstable_datafile_test.cc
Comment (haaawk) [inline test/boost/sstable_datafile_test.cc:6689]: The test is a bit simplistic but better than nothing. An extension would be to have more than one partition and only one of them having a partition tombstone.

Comment (kbr-) [reply test/boost/sstable_datafile_test.cc:6689]: done

File: sstables/sstables.cc
Comment (haaawk) [inline sstables/sstables.cc:1621]: std::move for a bool seems like overkill. I would remove std::move as it's useless noise.

File: sstables/kl/writer.cc
Comment (haaawk) [inline sstables/kl/writer.cc:666]: Why K/L sstables always have the new metadata set to false? Out of laziness or is there a reason?

**Review by haaawk — APPROVED** [review]
LGTM

File: sstables/sstables.cc
Comment (kbr-) [reply sstables/sstables.cc:1621]: without it the code doesn't compile ;)

File: sstables/kl/writer.cc
Comment (kbr-) [reply sstables/kl/writer.cc:666]: Laziness and I only really needed the metadata for >= md

File: test/boost/sstable_datafile_test.cc
Comment (raphaelsc) [inline test/boost/sstable_datafile_test.cc:6716]: I suggest you reload the sstable to force the metadata to be read from disk rather than served from memory as leftover by the writer.

File: sstables/sstables.cc
Comment (haaawk) [reply sstables/sstables.cc:1621]: wh00t!!! What's the error?

Comment (kbr-) [reply sstables/sstables.cc:1621]: It was something with boost's variant's `operator=` being defined for rvalue-reference only

File: sstables/mx/writer.cc
Comment (raphaelsc) [inline sstables/mx/writer.cc:611]: cannot_have_partition_tombstones is hard to parse to me. what about renaming it to contains_partition_tombstones and setting it to false by default?

File: sstables/kl/writer.cc
Comment (raphaelsc) [reply sstables/kl/writer.cc:666]: I'd like to understand this as well. if we set it to false by default, the metadata will be made available, and if it's set incorrectly, that could lead to a wrong decision down the road.

**Review by raphaelsc — CHANGES_REQUESTED** [review]

File: sstables/kl/writer.cc
Comment (haaawk) [reply sstables/kl/writer.cc:666]: I think it's safe because the metadata either says there's no partition tombstone or we don't know. Like a bloom filter.

Comment (haaawk) [reply sstables/kl/writer.cc:666]: false means we know nothing partition tombstones are there or not.

File: sstables/sstables.cc
Comment (raphaelsc) [reply sstables/sstables.cc:1621]: that's because disk_set_... defines set as follow:     template <TagType Tag, typename T>
    void set(T&& value) {
        data[Tag] = disk_tagged_union_member<TagType, Tag, T>{std::forward<T>(value)};
    }

alternatively, we could perfect forward cannot_... to it rather than moving

Comment (haaawk) [reply sstables/sstables.cc:1621]: nah. std::forward is even more noise than std::move.

File: sstables/kl/writer.cc
Comment (raphaelsc) [reply sstables/kl/writer.cc:666]: so perhaps we shouldn't even bother writing it for older versions

Comment (raphaelsc) [reply sstables/kl/writer.cc:666]: we do have the feature bit set in the scylla file however so that should be fine backward compatibility wise

File: sstables/types.hh
Comment (raphaelsc) [inline sstables/types.hh:501]: this doc would be be located at the top of cannot_have_partition_tombstones()

File: sstables/mx/writer.cc
Comment (raphaelsc) [inline sstables/mx/writer.cc:1007]: @kbr- could u pls explain to me why we don't need to track range tombstones as well? as part of https://github.com/scylladb/scylla/pull/7437 work. would we have a problem in 7437 if we have a range tombstone which is essentially a partition tombstone?

Comment (raphaelsc) [reply sstables/mx/writer.cc:1007]: In this commit https://github.com/scylladb/scylla/pull/7437/commits/34bb4d6da434d8cf55e8932c137da3466dc44abb, I found this comment:
    // 3. The sstables cannot have partition tombstones for the same reason as above. There
    //    is no way to check exactly this, but there is a way to check if the sstable has no tombstones
    //    of any type, so we use that. TWCS sstables will usually pass this condition.

But we aren't checking for range tombstones here, and I think they're important when doing the single-partition query filtering optimization.

Comment (kbr-) [reply sstables/mx/writer.cc:1007]: The comment is outdated. It turned out that we don't have a way that "sstable has no tombstones of any type", hence this PR.

Range tombstones are OK. Only partition tombstones are a problem since the new reader merger could encounter a partition tombstone in the middle of a partition query long after it returned a `partition_start` fragment.

Comment (kbr-) [reply sstables/mx/writer.cc:1007]: Range tombstones have a position in partition and are returned in order of their positions.

Partition tombstones are always returned as the first thing by mutation readers (in the `partition_start` fragment). However, the reader merger from #7437 is opening new mutation readers in the middle of a partition, after it already merged fragments returned from other readers. So imagine the following scenario: we're merging readers `r1` and `r2`:
1. the merger opens `r1`
2. `r1` returns `partition_start` with no tombstone, the merger returns it
3. `r1` returns a clustering row, the merger returns it
4. the merger opens `r2`
5. `r2` returns `partition_start` with a partition tombstone
6. ??? The merger cannot return it since it's illegal to return a `partition_start` twice. And it cannot merge it with the first `partition_start` since it already returned it.

The simplest way to deal with this problem is to ensure that there are no partition tombstones.

**Comment by kbr-** [comment]
Small fixes, see diff: https://github.com/scylladb/scylla/compare/4489df8441e2266c7f0ed74b650bcc90be666c31..4e90cac08ce3e80dad39eb579e9266fba02933f2

File: sstables/types.hh
Comment (kbr-) [reply sstables/types.hh]: done

File: test/boost/sstable_datafile_test.cc
Comment (kbr-) [reply test/boost/sstable_datafile_test.cc:6716]: done

File: sstables/kl/writer.cc
Comment (avikivity) [inline sstables/kl/writer.cc:666]: Better to use a bool_class, or even better a struct with a bool member. That naked `false` is less readable.

File: sstables/sstables.cc
Comment (avikivity) [inline sstables/sstables.cc:1621]: I find the name "cannot have" a little awkward. It's describing something already done, and immutable. More natural to say "has no".

Comment (avikivity) [reply sstables/sstables.cc:1621]: Or even, "has partition tombstones", to avoid the double negative.

No entry -> unknown
true -> has partition tombstones
false -> does not have partition tombstones

Comment (kbr-) [reply sstables/sstables.cc:1621]: but the negation to "has no" is "has", which would suggest that the flag being false means that the sstable has partition tombstones. But it merely means that it may have - we don't know if it does.

Comment (kbr-) [reply sstables/sstables.cc:1621]: https://github.com/scylladb/scylla/pull/7437#discussion_r521200478

File: sstables/mx/writer.cc
Comment (avikivity) [inline sstables/mx/writer.cc:612]: Maybe we should collect the lowest partition tombstone timestamp instead (and if there are none, it's the maximum timestamp). Maybe this can help with something, and has exactly the same collection cost.

@tgrabiec @raphaelsc maybe you can find a use for it.

File: sstables/sstables.cc
Comment (kbr-) [reply sstables/sstables.cc:1621]: Ok, maybe I'll just use an enum class `HasPartitionTombstones` with three options: `Yes`, `No`, `Unknown`

Comment (avikivity) [reply sstables/sstables.cc:1621]: There is a payload, a bool, no?

So if the section is missing, it's unknown, and if it exists, it's known, and the bool indicates what.

Comment (kbr-) [reply sstables/sstables.cc:1621]: I was worried about the API. If we define a function `has_partition_tombstones()`, it shouldn't return a `bool`, it should return a type with 3 possible values. On the other hand, if we define a function like `can_have_partition_tombstones()` or `cannot_have_partition_tombstones()` or `may_have_partition_tombstones()` or whatever, then it can return a `bool` and the semantics are consistent with the naming.

File: sstables/kl/writer.cc
Comment (kbr-) [reply sstables/kl/writer.cc]: Using `enum class` now.

**Comment by kbr-** [comment]
new version:
- `CannotHavePartitionTombstones` was replaced with `HasPartitionTombstones`. It is an enum class with three possible values: `Yes`, `No`, and `Unknown`.
- in `write_scylla_metadata` if the value of `has_partition_tombstones` is `Unknown`, we don't write it at all. This is used in the `kl` writer.
- `sstable::cannot_have_partition_tombstones` was renamed to `sstable::may_have_partition_tombstones` to avoid double negatives.

File: sstables/sstables.cc
Comment (kbr-) [reply sstables/sstables.cc]: check out the new version, now using enum class with 3 possible values.

File: sstables/mx/writer.cc
Comment (avikivity) [reply sstables/mx/writer.cc:612]: btw, the reason I am expanding the scope is that I'd like to make format changes as rare as we can.

Comment (raphaelsc) [reply sstables/mx/writer.cc:612]: in sst 3 format, there's min deletion time metadata to track if a sstable has tombstones of any sort, so it makes sense to have min partition tombstone deletion time to track for existence of partition tombstone.

Comment (kbr-) [reply sstables/mx/writer.cc:612]: > min deletion time metadata to track if a sstable has tombstones of any sort

Note that it's also updated on TTLs, hence it cannot be really used to track "real" tombstones. That's why I needed this new metadata.

Comment (kbr-) [reply sstables/mx/writer.cc:612]: > it makes sense to have min partition tombstone deletion time to track for existence of partition tombstone

But would it have an actual use? I could modify the PR to collect a timestamp instead of a flag but honestly I like it the way it is now :)

Comment (kbr-) [reply sstables/mx/writer.cc:612]: Btw. if we want to avoid format changes, maybe we could somehow make an extensible "statistics" metadata entry. Like `stats_metadata` but Scylla specific. However, `stats_metadata` is not extensible as far as I understand, and I don't know how we could make a nice extensible field (which can be appended to over time). I guess we could use a `disk_array` but that wouldn't be very elegant since it doesn't have type information. It would be best to represent the field as a `struct`. It probably could be done with a sufficient amount of template magic but I don't know if doing this as part of this PR wouldn't miss the point (I just want to finish #7437 :( ).

Comment (raphaelsc) [reply sstables/mx/writer.cc:612]: with the boolean, we gain simplicity and I like it more too. with the min timestamp, we have extra information that may or may not be required later on. however, extra information is useless if it will never be used. so we'd only go with the latter if anybody can think of where it could possibly be applied in the future.

Comment (kbr-) [reply sstables/mx/writer.cc:612]: > with the boolean, we gain simplicity and I like it more too. with the min timestamp, we have extra information that may or may not be required later on. however, extra information is useless if it will never be used. so we'd only go with the latter if anybody can think of where it could possibly be applied in the future.

also it's not the end of the world if we eventually end up adding a new field. It's inside Scylla-specific metadata anyway which works in a backwards-compatible way as far as I understand (does it?).

Comment (raphaelsc) [reply sstables/mx/writer.cc:612]: > > with the boolean, we gain simplicity and I like it more too. with the min timestamp, we have extra information that may or may not be required later on. however, extra information is useless if it will never be used. so we'd only go with the latter if anybody can think of where it could possibly be applied in the future.
> 
> also it's not the end of the world if we eventually end up adding a new field. It's inside Scylla-specific metadata anyway which works in a backwards-compatible way as far as I understand (does it?).

it does. scylla_metadata is capable of determining whether a metadata is available or not, so it's important to have a safe default value, like you did.

Comment (raphaelsc) [reply sstables/mx/writer.cc:612]: but please keep in mind that it's important to anticipate future usage of metadata to prevent scylla component from being eventually bloated. if we know that adding a bit more info to a metadata field can *very likely* be useful down the road, then go for it.

File: sstables/mx/writer.cc
Comment (raphaelsc) [reply sstables/mx/writer.cc]: thanks for the explanation!

**Review by raphaelsc — APPROVED** [review]
LGTM. loved the new version with enum for has_partition_tombstones

File: sstables/mx/writer.cc
Comment (avikivity) [inline sstables/mx/writer.cc:1007]: Wait. We wanted this new metadata to be sure to include an sstable that has a tombstone early. But the previous line (update_min_max_components) will already do that! Filtering based on clustering key will include it early, since the clustering range for this sstable will be empty.

Comment (kbr-) [reply sstables/mx/writer.cc:1007]: On the contrary, the clustering range will be the full range (`before_all_clustered`, `after_all_clustered`). This is what empty clustering key prefix denotes.

If we would set the sstable's clustering range to be empty on a partition tombstone, the current sstable reading functions would give incorrect results because they would not include sstables which have partition tombstones (the ck filter would filter them out).

If the range contains at least one mutation fragment in the partition, the range must include this fragment's position. Since partition tombstones are "everywhere" - they affect all other positions - if the sstable contains a partition tombstone, its range must be the full range.

Comment (kbr-) [reply sstables/mx/writer.cc:1007]: > We wanted this new metadata to be sure to include an sstable that has a tombstone early

This metadata is used to determine whether or not we should use the new TWCS-specific sstable-reader-merger. The merger only works correctly when there are no partition tombstones in the merged sstables. If there may be (which we check using this metadata), we use the old merger.

Comment (kbr-) [reply sstables/mx/writer.cc:1007]: I guess we could use the following method:
- use the new merger only if the clustering range of the sstable is not full.

Reason: it is safe to use the new merger if the clustering range is not full, because that means that it doesn't have partition tombstones.

However I'm not sure if this works for all sstables -- created by Cassandra or an older version of Scylla, i.e. if they enforce this condition -- so the safer condition would be:
- use the new merger only if the clustering range of the sstable is not full, the version is `>= md`, and the sstable has a Scylla component.

**Pull request closed without merging**
